### PR TITLE
Save entity methods

### PR DESCRIPTION
**Pull request by semseysandor**
Added `saveEntity()` methods from child class.

Mostly a simple copy, though there are some improvements, changes:
- moved abstract methods to the beginning of the class for better structure
- improved doc blocks for better documentation (mainly on the structure of array holding contact data)
- `saveEntity()` methods now return created entity ID
- renamed `saveEntity()` methods
- added `saveContactByEmail()`

### CODE REVIEW

File: CRM/Wrapi/Handler/Base.php [hunk 1]
```diff
@@ -333,21 +355,379 @@ protected function mapFieldsDateTimeISO8601(array $request_data, array $mapping)
...
+     *
+     * @param int $contact_id Contact ID
+     * @param array|null $address_data Address data
+     *   This contains the fields to save, it should be in a format which can be fed to civicrm_api4() calls.
+     *   See Api Explorer v4
+     *
+     *   Example:
+     *   $address_data = [
+     *     'location_type_id' => 1,
+     *     'mobile_provider_id' => 3,
```
Comment (akosgarai) [inline CRM/Wrapi/Handler/Base.php:645]: This example block seems to be connected to the phone not to the address.

Comment (semseysandor) [reply CRM/Wrapi/Handler/Base.php]: Fixed it. I've checked it some many times... :)

**Review by akosgarai — APPROVED** [review]
:+1:

**Comment by semseysandor** [comment]
I was wondering is it worth to check if supplied values for a given entity is actually different from the stores ones?
Then only update if there is any change.

I'm not sure how much we gain, but it is only a small coding effort, and could prevent an update query.
Instead there will always be a get.

Like:
```
$entity_data_stored = CRM_Wrapi_Actions_Get::entityData($entity_id);

if (is_null($entity_data_stored)) {
 throw new CRM_Core_Exception('Entity data not found in DB');
}

// Compare received vs. stored contact data
$data_changed = false;
foreach ($entity_data_new as $property => $value) {

  // If new is not empty and different --> update info
  if (!empty($value) && $value != $entity_data_stored[$property]) {
    $data_changed = true;
    break;
  }
}

// If there is a change in data --> update entity
if ($data_changed) {
  CRM_Wrapi_Actions_Update::entity($contact_id, $entity_data_new);
  $this->debug(sprintf('Contact updated (ID: %s)', $contact_id));
}

**Comment by akosgarai** [comment]
Checking the difference in the code is a good approach when the database is the bottleneck, and we need to decrease the number of the write db methods. The example code that you provide seems to be fine, but i think the deletion should also trigger an update.

```
if ((!empty($value) && $value != $entity_data_stored[$property]) || (empty($value) && !empty($entity_data_stored[$property])))) {
```

**Comment by semseysandor** [comment]
Cool I'll add an issue with this.

**Pull request merged**